src/pkcs11/core_pkcs11_test.c

```diff
@@ -0,0 +1,2647 @@
+                                            pxPublicKeyHandle );
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to find public key." );
+
+    if( *pxPrivateKeyHandle != CK_INVALID_HANDLE )
+    {
+        xTemplate.type = CKA_KEY_TYPE;
+        xTemplate.pValue = &xKeyType;
+        xTemplate.ulValueLen = sizeof( CK_KEY_TYPE );
+        xResult = pxGlobalFunctionList->C_GetAttributeValue( xGlobalSession, *pxPrivateKeyHandle, &xTemplate, 1 );
+        TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to find private key's type." );
+
+        if( xKeyType != CKK_EC )
+        {
+            xProvisionKeyNeeded = CK_TRUE;
+        }
+    }
+
+    if( *pxPrivateKeyHandle == CK_INVALID_HANDLE )
+    {
+        xProvisionKeyNeeded = CK_TRUE;
+    }
+
+    if( xProvisionKeyNeeded == CK_TRUE )
+    {
+        xResult = xProvisionGenerateKeyPairEC( xGlobalSession, ( uint8_t * ) PKCS11_TEST_LABEL_DEVICE_PRIVATE_KEY_FOR_TLS, ( uint8_t * ) PKCS11_TEST_LABEL_DEVICE_PUBLIC_KEY_FOR_TLS, pxPrivateKeyHandle, pxPublicKeyHandle );
+        TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to generate key pair." );
+        TEST_ASSERT_NOT_EQUAL_MESSAGE( CK_INVALID_HANDLE, *pxPrivateKeyHandle, "Invalid object handle returned for EC private key." );
+        TEST_ASSERT_NOT_EQUAL_MESSAGE( CK_INVALID_HANDLE, *pxPublicKeyHandle, "Invalid object handle returned for EC public key." );
+    }
+
+    xResult = xFindObjectWithLabelAndClass( xGlobalSession,
+                                            PKCS11_TEST_LABEL_DEVICE_CERTIFICATE_FOR_TLS,
+                                            sizeof( PKCS11_TEST_LABEL_DEVICE_CERTIFICATE_FOR_TLS ) - 1,
+                                            CKO_CERTIFICATE,
+                                            pxCertificateHandle );
+
+    /* NOTE: This certificate is for object storage and retrieval purposes only, and does not correspond to the key pair generated. */
+    if( *pxCertificateHandle == CK_INVALID_HANDLE )
+    {
+        xResult = xProvisionCertificate( xGlobalSession,
+                                         ( uint8_t * ) cValidECDSACertificate,
+                                         sizeof( cValidECDSACertificate ),
+                                         ( uint8_t * ) PKCS11_TEST_LABEL_DEVICE_CERTIFICATE_FOR_TLS,
+                                         pxCertificateHandle );
+        TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to create EC certificate." );
+        TEST_ASSERT_NOT_EQUAL_MESSAGE( 0, *pxPrivateKeyHandle, "Invalid object handle returned for EC certificate." );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+void prvProvisionEcTestCredentials( CK_OBJECT_HANDLE_PTR pxPrivateKeyHandle,
+                                    CK_OBJECT_HANDLE_PTR pxCertificateHandle,
+                                    CK_OBJECT_HANDLE_PTR pxPublicKeyHandle )
+{
+    #if ( PKCS11_TEST_IMPORT_PRIVATE_KEY_SUPPORT != 0 )

```

Comment:
There are some extra test assuming imported key. I would like to update the test so that all test cases will run on imported private key and key pair generation. Thank you for pointing out this problem.